### PR TITLE
Storybook doc formatting improvements

### PR DESCRIPTION
**Pull request by fredvisser**
# Pull Request

## 🤨 Rationale

We'd like to reflect richer and more visual content within Storybook. Particularly the Usage docs would benefit from more examples, along with clear `Do` and `Don't` examples.

## 👩‍💻 Implementation

- Created a handful of React components used for formatting usage content in Storybook (i.e. Frame, Container, Column, Do, Dont)
- Used the `@microsoft/fast-react-wrapper` to create React wrappers for all Nimble components
- Switched the `@storybook/html-webpack5` builder from babel to [SWC](https://storybook.js.org/docs/react/configure/babel#swc-alternative) to enable `support for the experimental syntax 'jsx'` without mucking with the babel config. (No issues noted thus far).
- Formatted the `Button.mdx` story better reflect the Figma source formatting using the new layout components.

Screenshot 2023-11-01 at 2 53 54 PM

Screenshot 2023-11-01 at 2 54 16 PM

## 🧪 Testing

- Manual review of the `Button.mdx` story

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

**Comment by fredvisser** [comment]
@NIbokeefe - Particularly interested in your feedback on this change. This PR creates a handful of simple layout components that can be reused within Storybook documentation. I've updated the [Button usage docs](https://60e89457a987cf003efc0a5b-hepeosrekr.chromatic.com/?path=/docs/components-button--docs) to reflect a typical use case.

I'd appreciate any and all feedback – particularly on visual design (colors, margins, etc.) and if these layout patterns are useful, or if you have other ideas for layout patterns we could reuse within usage docs to visually communicate best practices. Thanks!

**Comment by NIbokeefe** [comment]
@fredvisser Thank you Fred, at first glance this looks really nice. I  only had a minute to overview it right now, but will dig into it later.

One thing I noticed you could look into immediately is the Accent/Primary button section should promote the use of the accent - outline version as the first instance to use.  Then follow up with the block-accent version.  Similar to the order of the standard buttons.

**Comment by NIbokeefe** [comment]
> @NIbokeefe - Take a look at the [latest changes](https://60e89457a987cf003efc0a5b-ymddauwptm.chromatic.com/iframe.html?viewMode=docs&id=components-button--docs#accent-button) to see if I've mapped your feedback correctly.
> 
> @jattasNI - Could you take an initial look at this PR?

@fredvisser That solved the issue, looks good.  Thanks.

File: change/@ni-nimble-components-be528419-2f60-4d4b-b6f0-a8b8e1e79bae.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (jattasNI) [inline change/@ni-nimble-components-be528419-2f60-4d4b-b6f0-a8b8e1e79bae.json:1]: Once we have agreement on the pattern for adding a react wrapper, we should update README and CONTRIBUTING and issue templates to remind us to create it (and explain how to create it) when building a new component.

File: packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/* eslint-disable import/no-extraneous-dependencies */
+/* eslint-disable @typescript-eslint/naming-convention */
+import * as React from 'react';
+import { provideReactWrapper } from '@microsoft/fast-react-wrapper';
+import { AnchorMenuItem } from '..';
+
+const { wrap } = provideReactWrapper(React);
+
+export const NimbleButton = wrap(AnchorMenuItem);
```
Comment (jattasNI) [inline packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx:9]: I'm guessing every component shouldn't be exported as `NimbleButton`?

I'm debating between `NimbleAnchorMenuItem` and something with React / Storybook in its name like `NimbleStorybookAnchorMenuItem` or `AnchorMenuItemReactWrapper`. My gut is that we probably want to avoid any chance of confusing these for the web components so it's worth including something like that, but I don't currently have a strong opinion.

File: packages/nimble-components/src/button/index.ts [hunk 3]
```diff
@@ -58,7 +58,7 @@ export class Button
  * Generates HTML Element: \<nimble-button\>
  *
  */
-const nimbleButton = Button.compose<ButtonOptions>({
+export const nimbleButton = Button.compose<ButtonOptions>({
```
Comment (jattasNI) [inline packages/nimble-components/src/button/index.ts:61]: Is this change still needed? Why was it only needed for the button?

File: packages/nimble-components/.storybook/blocks/StoryLayout.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
```
Comment (jattasNI) [inline packages/nimble-components/.storybook/blocks/StoryLayout.tsx:1]: Could be useful to add some docs for this infrastructure:
1. a short README in the `blocks` folder explaining what goes there
2. a JSDoc comment on each block (at least the top level ones) explaining what it's used for and maybe even a code snippet showing its usage

File: packages/nimble-components/src/icon-base/tests/icons.react.tsx [hunk 5]
```diff
@@ -0,0 +1,363 @@
+/* eslint-disable import/no-extraneous-dependencies */
+/* eslint-disable @typescript-eslint/naming-convention */
+import * as React from 'react';
```
Comment (jattasNI) [inline packages/nimble-components/src/icon-base/tests/icons.react.tsx:3]: Don't forget to update the instructions for adding a new icon to mention also updating this file.

File: packages/nimble-components/src/button/tests/button.react.tsx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (jattasNI) [inline packages/nimble-components/src/button/tests/button.react.tsx:1]: I'd vote to add configuration to eslintrc rather than inline disables in each file. You should be able to do that with an `overrides` rule targeting `*.react.tsx`. That would also allow you to add a comment justifying the exceptions.

File: change/@ni-nimble-components-be528419-2f60-4d4b-b6f0-a8b8e1e79bae.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+  "type": "none",
```
Comment (jattasNI) [inline change/@ni-nimble-components-be528419-2f60-4d4b-b6f0-a8b8e1e79bae.json:2]: It'd be good to make sure the react wrappers aren't ending up in the built `nimble-components` package. You could do that by manual inspection of the built package (it should be available as a pipeline artifact). I'm not sure what we could do to make sure they aren't accidentally included later. Maybe better alarming when the package size increases?

Unless someone else has ideas I'm ok with one time manual verification for now. Longer term is the idea they'd go in a completely separate package? But that's too much work because it leads to a circular dependency? (storybook in nimble-components) -> (react components) -> (components in nimble components).

Comment (fredvisser) [reply change/@ni-nimble-components-be528419-2f60-4d4b-b6f0-a8b8e1e79bae.json:2]: I've confirmed that the react wrappers aren't ending up in the built `nimble-components` package in the current setup. The rollup config is targeting the components `index.ts` file, so I don't expect the react wrapper to easily get sucked into the package. 🤷‍♂️

I tried to autogenerate the wrappers, but failed to make much progress by extending the rollup config. I too, am OK with the manual approach for now… it's literally 10 seconds of work for a new component, and shouldn't change much.

File: packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx (hunk 2 above)
Comment (fredvisser) [reply packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx:9]: I went with `NimbleStorybook` so that it's clear within the MDX files that it's a Nimble component and to set expectations that these are not general-purpose React components - 👍

File: packages/nimble-components/src/icon-base/tests/icons.react.tsx (hunk 5 above)
Comment (rajsite) [reply packages/nimble-components/src/icon-base/tests/icons.react.tsx:1]: We definitely should be modifying the icon build script instead https://github.com/ni/nimble/blob/main/packages/nimble-components/build/generate-icons/source/index.js

File: packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import * as React from 'react';
+import { provideReactWrapper } from '@microsoft/fast-react-wrapper';
+import { AnchorMenuItem } from '..';
+
+const { wrap } = provideReactWrapper(React);
```
Comment (rajsite) [inline packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx:5]: I think we should avoid importing React and invoking provideReactWrapper in every file. Should just be done in one spot and exported file should just be
```ts
import { AnchorMenuItem } from '..';
import { wrap } from 'somewhere-probably-in-the-storybook-utils-folder';
export const NimbleAnchorMenuItem = wrap(AnchorMenuItem);
```

File: packages/nimble-components/src/button/tests/button.mdx [hunk 9]
```diff
@@ -28,30 +29,104 @@ These appearances have the default styling of the nimble-button. Each should be
...
-Ghost is the default style and should be the first consideration for use.
-
-Use as the default and standard option to create a clean airy and open UI feel. Ghost buttons fit comfortably in tight spaces and help control the visual density of the UI.
-Be careful when using that the surrounding context does not cause this button to be confused for emphasized body text, tabs or a standalone links.
-
-Use in combination with a primary outline and primary block buttons to create a hierarchy of importance. There is no primary ghost button.
+<Frame>
+    <Container>
+        <Column stylingClass="controls">
+        <NimbleStorybookButton appearance="ghost">{`Ghost Button`}</NimbleStorybookButton>
```
[the remark this answers is not in the document]
Comment (fredvisser) [reply packages/nimble-components/src/button/tests/button.mdx:35]: I changed it in response to the feedback here - https://github.com/ni/nimble/pull/1637#discussion_r1408547237 - but don't have a strong opinion either way.

[the remark this answers is not in the document]
Comment (fredvisser) [reply packages/nimble-components/src/button/tests/button.mdx]: Changed it back to `NimbleButton`

File: packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx (hunk 2 above)
Comment (fredvisser) [reply packages/nimble-components/src/anchor-menu-item/tests/anchor-menu-item.react.tsx]: Changed it back to `NimbleButton` per https://github.com/ni/nimble/pull/1637#discussion_r1418248363

File: packages/nimble-components/src/dialog/tests/dialog.react.tsx [hunk 10]
```diff
@@ -0,0 +1,6 @@
+// template-folder-name -> dialog.js
+
```
Comment (rajsite) [inline packages/nimble-components/src/dialog/tests/dialog.react.tsx:2]: What does this mean? can we remove it?

File: packages/nimble-components/src/button/tests/button.mdx [hunk 11]
```diff
@@ -5,16 +5,17 @@ import {
     Primary,
     Stories,
     Title,
-    Description
+    Description,
+    Story
 } from '@storybook/blocks';
 import ContentHiddenDocs from '../../patterns/button/tests/content-hidden-docs.mdx';
+import { NimbleButton } from './button.react';
+import { NimbleIconKey } from '../../icons/all-icons.react';
```
Comment (rajsite) [inline packages/nimble-components/src/button/tests/button.mdx:13]: Can we make sure that goes into a folder named /icons/tests/all-icons.react

File: packages/nimble-components/build/generate-icons/source/index.js [hunk 12]
```diff
@@ -70,10 +75,21 @@ export const ${tagName} = DesignSystem.tagFor(${className});
...
 
 const allIconsFilePath = path.resolve(iconsDirectory, 'all-icons.ts');
 console.log('Writing all-icons file');
 fs.writeFileSync(allIconsFilePath, allIconsFileContents, { encoding: 'utf-8' });
 console.log('Finished writing all-icons file');
+
+allIconsReactFileImports = allIconsReactFileImports.concat('} from \'./all-icons\';\n');
+const allIconsReactFileContents = allIconsReactFileImports.concat(allIconsReactFileExports);
+
+const allIconsReactFilePath = path.resolve(iconsDirectory, 'all-icons.react.tsx');
```
Comment (rajsite) [inline packages/nimble-components/build/generate-icons/source/index.js:92]: I don't think it would be a huge effort to emit these as `/icons/tests/nimble-icon.react.tsx`
That would follow our other icon patterns and prevent two issues:
- A test importing a single icon will import every icon. Each story is a support output now so importing any icon will make them all larger now (possibly)
- Modifying / adding an icon will cause every snapshot to be remade since pretty much every story will have a dependency on all-icons.

I think it's worthwhile to emit the separate icon files for consistency and avoid the (possible) downsides.

File: packages/nimble-components/src/icon-base/tests/icons.react.tsx (hunk 5 above)
Comment (fredvisser) [reply packages/nimble-components/src/icon-base/tests/icons.react.tsx]: Added, then removed, the instructions as the script creates the React wrappers for icons automatically now.

File: packages/nimble-components/build/generate-icons/source/index.js (hunk 12 above)
Comment (fredvisser) [reply packages/nimble-components/build/generate-icons/source/index.js]: Sounds good - we now have the icon react wrapper generated like so – 

`packages/nimble-components/src/icons/tests/add.react.tsx`

```jsx
// AUTO-GENERATED FILE - DO NOT EDIT DIRECTLY
// See generation source in nimble-components/build/generate-icons

import { wrap } from '../../utilities/tests/react-wrapper';
import { IconAdd } from '../add';

export const NimbleIconAdd = wrap(IconAdd);
```

File: packages/nimble-components/.storybook/blocks/assets/Check.tsx [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+interface CheckProps extends React.SVGAttributes<SVGElement> {
+  size?: string | number;
+}
+const Check: React.FC<CheckProps> = ({ size, ...props }) => (
+    <svg width={size || '32'} height={ size || '32'} viewBox="0 0 32 32" fill="none"
+        xmlns="http://www.w3.org/2000/svg">
+        <path d="M28 9.13502L11.636 25.501L4 17.863L7.272 14.599L11.772 19.099L24.864 5.99902L28 9.13502Z" fill="#00C12B"/>
+    </svg>
```
Comment (rajsite) [inline packages/nimble-components/.storybook/blocks/assets/Check.tsx:9]: We already have a lovely check, maybe we can use NimbleIconCheck here instead?

Comment (rajsite) [reply packages/nimble-components/.storybook/blocks/assets/Check.tsx:9]: Or I guess use it in Do directly, may need need this file

File: packages/nimble-components/.storybook/blocks/assets/Exclamation.tsx [hunk 14]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+interface ExclamationProps extends React.SVGAttributes<SVGElement> {
+  size?: string | number;
+}
+const Exclamation: React.FC<ExclamationProps> = ({ size, ...props }) => (
+  <svg width={size || '32'} height={ size || '32'} viewBox="0 0 32 32" fill="none"
```
[the remark this answers is not in the document]
Comment (rajsite) [reply packages/nimble-components/.storybook/blocks/assets/Exclamation.tsx:6]: Same as above, could use the nimble icon in Dont directly

File: packages/nimble-components/.storybook/blocks/assets/Check.tsx (hunk 13 above)
Comment (fredvisser) [reply packages/nimble-components/.storybook/blocks/assets/Check.tsx]: Good call - removed - 👍

File: packages/nimble-components/src/button/tests/button.mdx [hunk 15]
```diff
@@ -28,30 +29,104 @@ These appearances have the default styling of the nimble-button. Each should be
...
+        <NimbleButton appearance="ghost">{`Ghost Button`}</NimbleButton>
+        <NimbleButton appearance="ghost" content-hidden>
+            <NimbleIconKey slot="start"></NimbleIconKey>
+            {`Ghost Button`}
+        </NimbleButton>
+
+        </Column>
+        <Column>
+            <ul>
+                <li>Ghost is the default style and should be the first consideration for use.</li>
```
Comment (jattasNI) [inline packages/nimble-components/src/button/tests/button.mdx:44]: nit: grammar
```suggestion
                <li>Ghost is the default appearance and should be the first considered for use.</li>
```

File: packages/nimble-components/src/button/tests/button.mdx [hunk 16]
```diff
@@ -62,32 +137,82 @@ Button appearance variants are mainly used when a button needs be distinguished
...
+            <NimbleButton appearance="ghost">{`Ghost Button`}</NimbleButton>
+            <NimbleButton appearance="outline" appearance-variant="primary">
+            {`Primary Button`}
+            </NimbleButton>
+            <Divider/>
+            <NimbleButton appearance="ghost">{`Ghost Button`}</NimbleButton>
+            <NimbleButton appearance="ghost">{`Ghost Button`}</NimbleButton>
+            <NimbleButton appearance="block" appearance-variant="primary">
+                {`Primary Button`}
+            </NimbleButton>
```
Comment (jattasNI) [inline packages/nimble-components/src/button/tests/button.mdx:196]:
```suggestion
            <NimbleButton appearance="outline" appearance-variant="primary">
            {`Outline Primary Button`}
            </NimbleButton>
            <Divider/>
            <NimbleButton appearance="ghost">{`Ghost Button`}</NimbleButton>
            <NimbleButton appearance="ghost">{`Ghost Button`}</NimbleButton>
            <NimbleButton appearance="block" appearance-variant="primary">
                {`Block Primary Button`}
            </NimbleButton>
```

File: packages/nimble-components/.storybook/blocks/README.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+-   Use CSS [grid-template-columns](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns) for the config prop to specify the number of columns and their widths.
+
+### Column
+
+-   Renders a column for use with `<Container>`.
+-   If you need to put Nimble components in a column, use the stylingClass="controls" prop to apply the correct margins.
+
+### Divider
+
+-   Renders a simple divider to vertically separate the contents of a `<Column>`.
```
Comment (jattasNI) [inline packages/nimble-components/.storybook/blocks/README.md:25]:
```suggestion
-   Renders a simple horizontal divider to vertically separate the contents of a `<Column>`.
```

File: packages/nimble-components/.storybook/blocks/README.md [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+### Do
+
+-   Renders a "Do" section for MDX documentation.
+
+### Dont
+
+-   Renders a "Dont" section for MDX documentation.
```
Comment (jattasNI) [inline packages/nimble-components/.storybook/blocks/README.md:33]:
```suggestion
-   Renders a "Don't" section for MDX documentation.
```

**Review by jattasNI — APPROVED** [review]

File: packages/nimble-components/src/button/tests/button.mdx [hunk 19]
```diff
@@ -5,16 +5,17 @@ import {
...
 } from '@storybook/blocks';
 import ContentHiddenDocs from '../../patterns/button/tests/content-hidden-docs.mdx';
+import { NimbleButton } from './button.react';
+import { NimbleIconKey } from '../../icons/tests/key.react';
 import * as buttonStories from './button.stories';
 
 <Meta of={buttonStories} />
-
 <Title of={buttonStories} />
 <Description of={buttonStories} />
```
[the remark this answers is not in the document]
Comment (fredvisser) [reply packages/nimble-components/src/button/tests/button.mdx:18]: I pulled the "A common convention…" sentence out for now, but will continue the effort in another PR to standardize the usage docs.
